gh-99633: Add context manager support to `contextvars.Context`

rhansen opened the pull request:
<!-- gh-issue-number: gh-99633 -->
* Issue: gh-99633
<!-- /gh-issue-number -->

sobolevn commented on Python/context.c:
```diff
@@ -558,6 +557,47 @@ context_tp_contains(PyContext *self, PyObject *key)
+/*[clinic input]
+_contextvars.Context.__enter__
+
+Context manager enter.
```
This will go into `__doc__` of these methods. Probably it can be something more useful. Like: a short sentence about when it should be used, maybe even mentioning `.run` alternative.

rhansen replied:
Done.

sobolevn commented on Python/context.c:
```diff
@@ -558,6 +557,47 @@ context_tp_contains(PyContext *self, PyObject *key)
+    if (_PyContext_Exit(ts, (PyObject *)self)) {
+        return NULL;
+    }
+    Py_RETURN_FALSE;
```
```suggestion
    Py_RETURN_NONE;
```

rhansen replied:
Done.

sobolevn commented on Python/context.c:
```diff
@@ -558,6 +557,47 @@ context_tp_contains(PyContext *self, PyObject *key)
+    if (_PyContext_Enter(ts, (PyObject *)self)) {
+        return NULL;
+    }
+    return Py_NewRef(self);
```
Reviewing note: `self` object is incremented twice. Here and in `_PyContext_Enter`: `ts->context = Py_NewRef(ctx);`

rhansen replied:
Yes.  The refcount added in `_PyContext_Enter` is subtracted in `_PyContext_Exit` by the (oddly named) `Py_SETREF` macro.

Both increments are necessary:  one is to keep the Context alive while it is entered, and one is for the `with` statement's `as` binding.

picnixz replied:
Should we add a comment for people reading the code?

rhansen replied:
Done.

sobolevn commented on Lib/test/test_context.py:
```diff
@@ -360,6 +361,62 @@ def sub(num):
+                with ctx:
+                    pass
+
+    def test_context_manager_4(self):
```
Let's add more semantic names to these tests.
For example, this one can be: `test_run_inside_context_raises`.

However, this is the naming style that is used in `test_context` right now.

rhansen replied:
Done.  I also split the first test into three to clarify intentions.

sobolevn commented on Lib/test/test_context.py:
```diff
@@ -360,6 +361,62 @@ def sub(num):
+        self.assertEqual(cvar.get(), 'initial')
+        ctx = contextvars.copy_context()
+        with ctx as tmp:
+            self.assertIs(tmp, ctx)
```
Do we really need to return `tmp` from `__enter__` here? We now have two identical identities with different names. It might confuse users.

rhansen replied:
Addressed by splitting the `assertIs` check to its own test case.

sobolevn commented on Lib/test/test_context.py:
```diff
@@ -360,6 +361,62 @@ def sub(num):
+    @threading_helper.requires_working_threading()
+    def test_context_manager_2(self):
+        ctx = contextvars.copy_context()
+        thread = threading.Thread(target=ctx.__enter__)
```
I think it won't hurt to add a thread-safety test, something like `test_context_threads_1`. How hard would it be?

rhansen replied:
I don't quite understand what `test_context_threads_1` is meant to test.  Maybe it's testing that each thread automatically gets its own Context instance?  If threads didn't automatically get their own context then that test would fail with high probability, I think.

If that test is meant to test internal state data races then I don't understand how it accomplishes that goal, as sleeping for even 1ms significantly drops the probability of losing a data race.

It would be great to write a test that guarantees that concurrent `Context.__enter__()` and/or `Context.run()` calls on the same Context object will result in exactly one successful call, and `RuntimeError` exceptions for the other calls, without any internal state corruption due to data races.  I'm not sure how to write such a test.

However, I'm not sure it's worth adding such a test in this PR:
  * The possibility of concurrent enters is not new to this PR (one can call `Context.run()` concurrently today).
  * This PR doesn't make concurrent enters more likely.
  * Users are unlikely to encounter a race because Context objects are not meant to be used across threads.  (There is no way to hand off ownership of an entered Context to another thread, and exiting a Context from a different thread will raise.)
  * I think the test would fail because `_PyContext_Enter` does not atomically check and set `ctx->ctx_entered`.  Fixing that feels out of scope for this PR.

It might be worth addressing thread safety in another PR, but as far as I know nobody has complained so far, so maybe it's better to just leave it alone.

sobolevn reviewed (commented):
Thank you! Looks interesting :)

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
    Context implements the :class:`collections.abc.Mapping` interface.
 
+   .. versionadded:: 3.14
+      Context implements the :term:`!context management protocol` (:pep:`343`).
```
```suggestion
      Added support for the :term:`context management protocol`.
```

Let's reduce the number of links.

rhansen replied:
"Context management protocol" is not defined in the glossary, which is why I linked to the PEP instead.

I could do ``:term:`context manager` protocol`` instead, although that terminology doesn't exactly match "context management protocol" used in the PEP.

Or I could add "context management protocol" to the glossary.

WDYT?

cjw296 replied:
Yep, let's do that.

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+      :meth:`__enter__` method enters the Context and returns *self*, so the
+      value bound to the identifier given in the :keyword:`with` statement's
+      :keyword:`!as` clause (if present) is the Context object itself.  The
+      :meth:`__exit__` method exits the Context.
```
```suggestion
```

Other classes don't seem to document those kind of methods. Since they don't do much more than what you already explained, we can simplify the docs (but keep the examples).

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+         with ctx:
+             assert var.get() == "updated"
+
+      .. versionadded:: 3.14
```
```suggestion
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+
+         # Re-entering ctx restores the updated value of var.
+         with ctx:
+             assert var.get() == "updated"
```
I'd suggest using the same formulation as in `run()` where we don't use `assert` but rather comments.

rhansen replied:
Note the use of `.. testcode::`.  I want this example to be testable to prevent the implementation from drifting away from the documentation.  Would you prefer I use `print` and `.. testoutput::`?  I personally find `assert` easier to read.

cjw296 replied:
Is that test code actually checked, though? I guess put differently: can you point me at the CPython CI that's actually executing this test code and checking it is as expected?

picnixz replied:
That would be the doctest CI job. This is a sphinx directive for running doctest: https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#directive-testcode.

---

> Note the use of .. testcode::. I want this example to be testable to prevent the implementation from drifting away from the documentation. Would you prefer I use print and .. testoutput::? I personally find assert easier to read.

I think print() could be preferrable so that users, if they C/C the code, would also see the output directly. With a combination of `.. testoutput` that would be enough.

You can also change the rst for run() so that it also uses `testcode/testoutput`. Alternatively, you can use the `doctest` directive directly.

rhansen replied:
I updated the `run` documentation to also use `testcode`/`testoutput`.

picnixz commented on Lib/test/test_context.py:
```diff
@@ -361,6 +362,74 @@ def sub(num):
         self.assertEqual(results, list(range(10)))
 
+    @isolated_context
+    def test_context_manager_works(self):
```
```suggestion
    def test_context_manager(self):
```

picnixz commented on Lib/test/test_context.py:
```diff
@@ -361,6 +362,74 @@ def sub(num):
+        with self.assertRaises(RuntimeError):
+            ctx.__exit__(None, None, None)
+
+    def test_context_manager_rejects_recursive_enter_mgr_then_mgr(self):
```
```suggestion
    def test_context_manager_is_not_reentrant(self):
```

Re-entrancy in this case would mean that I can nest `with` statements, right? or is the term not adequate?

picnixz commented on Lib/test/test_context.py:
```diff
@@ -361,6 +362,74 @@ def sub(num):
+    def test_context_manager_rejects_recursive_enter_mgr_then_run(self):
+        with contextvars.copy_context() as ctx:
+            with self.assertRaises(RuntimeError):
+                ctx.run(lambda: None)
```
I think this test can be put in the above test, together with using `self.subTest` to indicate what kind of behaviour you're testing.

picnixz commented on Lib/test/test_context.py:
```diff
@@ -361,6 +362,74 @@ def sub(num):
+            with self.assertRaises(RuntimeError):
+                ctx.run(lambda: None)
+
+    def test_context_manager_rejects_recursive_enter_run_then_mgr(self):
```
Ditto.

picnixz commented on Lib/test/test_context.py:
```diff
@@ -361,6 +362,74 @@ def sub(num):
+            with contextvars.copy_context():
+                with self.assertRaises(RuntimeError):
+                    ctx.__exit__(None, None, None)
+
```
```suggestion
        with contextvars.copy_context() as outer_ctx:
            with contextvars.copy_context() as inner_ctx:
                self.assertIsNot(outer_ctx, inner_ctx)
                with self.assertRaises(RuntimeError):
                    outer_ctx.__exit__(None, None, None)
```

picnixz commented on Misc/NEWS.d/next/Library/2022-11-21-01-24-46.gh-issue-99633.vhrNRe.rst:
```diff
@@ -0,0 +1 @@
+Add :term:`context manager` methods to :class:`contextvars.Context`.
```
```suggestion
Added support for the :term:`context management protocol` to
:class:`contextvars.Context` objects. Patch by Richard Hansen.
```

The feature is major enough (IMO) to have an additional What's New entry (not everyone reads the full changelog and this is a nice feature).

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+
+Automatically called by the 'with' statement.  Using the Context object as a
+context manager is an alternative to calling the Context.run() method.  Example
+usage:
```
```suggestion
Automatically called by the 'with' statement.  Using the Context object as a
context manager is an alternative to calling the Context.run() method.

Example:
```

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+    with contextvars.copy_context():
+        # The current Context is a new copy of the previous Context.  Updating a
+        # context variable inside this 'with' statement only affects the new
+        # copy.
```
```suggestion
        # Changes to context variables will be rolled back upon
        # exiting the `with` statement.
```

Maybe simplify the wording a bit?

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+
+    # Now that the 'with' statement is done executing, the value of the 'var'
+    # context variable has reverted back to its value before the 'with'
+    # statement.
```
```suggestion
    # The context variable value has been rolled back.
```

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+    # Now that the 'with' statement is done executing, the value of the 'var'
+    # context variable has reverted back to its value before the 'with'
+    # statement.
+    assert var.get() == 'initial'
```
```suggestion
    print(var.get())  # 'initial'
```

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+        # context variable inside this 'with' statement only affects the new
+        # copy.
+        var.set('updated')
+        do_something_interesting()
```
```suggestion
        print(var.get())  # 'updated'
```

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+    PyThreadState *ts = _PyThreadState_GET();
+    if (_PyContext_Exit(ts, (PyObject *)self)) {
+        return NULL;
+    }
```
```suggestion
	if (PyContext_Exit((PyObject *)self) < 0) {
        return NULL;
    }
```

And if you want something even more concise, you can have something like:

```C
return PyContext_Exit((PyObject *)self) < 0 ? NULL : Py_None;
```

picnixz commented on Python/context.c:
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
+    PyThreadState *ts = _PyThreadState_GET();
+    if (_PyContext_Enter(ts, (PyObject *)self)) {
+        return NULL;
+    }
```
```suggestion
	if (PyContext_Enter((PyObject *)self) < 0) {
        return NULL;
    }
```

And the concise form:

```C
return PyContext_Enter((PyObject *)self) < 0 ? NULL : Py_NewRef(self);
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+   Each thread has its own effective stack of :class:`~contextvars.Context`
+   objects.  The *current Context* is the Context object at the top of the
+   current thread's stack.  All Context objects in the stacks are considered to
+   be *entered*.  
```
```suggestion
   Each thread has its own effective stack of :class:`Context` objects.
   The *current context* is the :class:`!Context` object at the top of
   the current thread's stack.  All :class:`!Context` objects in the
   stacks are considered to be *entered*.
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+   making it the current Context.  *Exiting* from the current Context, either by
+   returning from the callback passed to :meth:`Context.run` or by exiting the
+   :keyword:`with` statement suite, pops the Context off of the top of the
+   stack, restoring the current Context to what it was before.
```
```suggestion
   *Entering* a context, either by calling the :meth:`~Context.run` method
   or using the context object as a :term:`context manager`, pushes the
   context onto the top of the current thread's stack, making it the
   current context.

   *Exiting* from the current context, either by returning from the callback
   passed to :meth:`~Context.run` or by exiting the :keyword:`with` statement
   suite, pops the contextoff of the top of the stack, restoring the current
   context to what it was before.
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+   :keyword:`with` statement suite, pops the Context off of the top of the
+   stack, restoring the current Context to what it was before.
+
+   Because each thread has its own Context stack, :class:`ContextVar` objects
```
```suggestion
   Since each thread has its own context stack, :class:`ContextVar` objects
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+   behave in a similar fashion to :func:`threading.local` when values are
+   assigned in different threads.
+
+   Attempting to do either of the following will raise a :exc:`RuntimeError`:
```
```suggestion
   Attempting to do either of the following raises a :exc:`RuntimeError`:
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+
+   * Entering an already entered Context.  (This includes Contexts entered in
+     other threads.)
+   * Exiting from a Context that is not the current Context.
```
```suggestion
   * Entering an already entered context, including contexts entered in
     other threads.
   * Exiting from a context that is not the current context.
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+     other threads.)
+   * Exiting from a Context that is not the current Context.
+
+   After exiting a Context, it can later be re-entered (from any thread).
```
```suggestion
   After exiting a context, it can later be re-entered (from any thread).
```

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,89 @@ Manual Context Management
+   method returns the value associated with the current Context.  Thus, exiting
+   a Context effectively reverts any changes made to context variables while the
+   Context was entered.  (If desired, the values can be restored by re-entering
+   the Context.)
```
```suggestion
   Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
   method are recorded in the current context. The :meth:`ContextVar.get`
   method returns the value associated with the current context. Exiting
   a context effectively reverts any changes made to context variables while
   the context was entered (if needed, the values can be restored by re-entering
   the context).
```

1st1 reviewed (changes requested):
I have a feeling this will not work at all with async/await. If that's the case then this API is a no-go. 

Please add a series of async/await tests with multiple context switches and contexts to show that this primitive is composable.

picnixz commented on Doc/library/contextvars.rst:
```diff
@@ -144,21 +144,79 @@ Manual Context Management
+      The value bound to the identifier given in the :keyword:`with` statement's
+      :keyword:`!as` clause (if present) is the :class:`!Context` object itself.
+
+      Example:
```
The example should be out of the `.. versionchanged` section, ideally before (we should use `versionchanged` instead of `versionadded` since the latter is only when the object is being added, not when features are added).

rhansen committed:
gh-99633: Add context manager support to `contextvars.Context`

rhansen committed:
gh-99633: _contextvars: Change private types to reduce casting

Change `PyObject *` to/from `PyContext *` to reduce the amount of
casting and improve readability.

rhansen committed:
gh-99633: _contextvars: Clarify that the ref is stolen, not borrowed

Improve readability by moving destination assignment next to source
reset, and comment that the ref is stolen.

rhansen committed:
gh-99633: _contextvars: Enter the thread's initial context

No public API provides access to the current context yet (only a new
copy), however:

  * This is good defensive practice.
  * This improves code readability.
  * Context watchers are now notified about the initial context.
  * A planned future commit will make it possible for users to access
    the thread's initial context object.  Without this change, users
    would be able to enter the context a second time, causing a cycle
    in the context stack.

rhansen committed:
gh-99633: New generator/coroutine `_context` property

Add a new `_context` property to generator (and coroutine) objects to
get/set the "current context" that is observed by (and only by) the
generator and the functions it calls.

When `generator._context` is set to `None` (the default), the
generator is called a "dependent generator".  It behaves the same as
it always has: the "current context" observed by the generator is the
thread's context.  This means that the observed context can change
arbitrarily during a `yield`; the generator *depends* on the sender to
enter the appropriate context before it calls `generator.send`.

When `generator._context` is set to a `contextvars.Context` object,
the generator is called an "independent generator".  It acts more like
a separate thread with its own independent context stack.  The value
of `_context` is the head of that independent stack.  Whenever the
generator starts or resumes execution (via `generator.send`), the
current context temporarily becomes the generator's associated
context.  When the generator yields, returns, or propagates an
exception, the current context reverts back to what it was before.
The generator's context is *independent* from the sender's context.

If an independent generator calls `contextvars.Context.run`, then the
value of the `_context` property will (temporarily) change to the
newly entered context.

If an independent generator sends a value to a second independent
generator, the second generator's context will shadow the first
generator's context until the second generator returns or yields.

The `generator._context` property is private for now until experience
and feedback is collected.  Nothing is using this yet, but that will
change in future commits.

Motivations for this change:

  * First, this change makes it possible for a future commit to add
    context manager support to `contextvars.Context`.  A `yield` after
    entering a context causes execution to leave the generator with a
    different context at the top of the context stack than when
    execution started.  Swapping contexts in and out when execution
    suspends and resumes can only be done by the generator itself.

  * Second, this paves the way for a public API that will enable
    developers to guarantee that the context remains consistent
    throughout a generator's execution.  Right now the context can
    change arbitrarily during a `yield`, which can lead to subtle bugs
    that are difficult to root cause.  (Coroutines run by an asyncio
    event loop do not suffer from this same problem because asyncio
    manually sets the context each time it executes a step of an
    asynchronous function.  See the call to `contextvars.Context.run`
    in `asyncio.Handle._run`.)

  * Finally, this makes it possible to move the responsibility for
    activating an async coroutine's context from the event loop to the
    coroutine, where it more naturally belongs (alongside the rest of
    the execution state such as local variable bindings and the
    instruction pointer).  This ensures consistent behavior between
    different event loop implementations.

Example:

```python
import contextvars

cvar = contextvars.ContextVar('cvar', default='initial')

def make_generator():
    yield cvar.get()
    yield cvar.get()
    yield cvar.get()
    yield cvar.get()
    cvar.set('updated by generator')
    yield cvar.get()

gen = make_generator()
print('1.', next(gen))

def callback():
    cvar.set('updated by callback')
    print('2.', next(gen))

contextvars.copy_context().run(callback)
print('3.', next(gen))
cvar.set('updated at top level')
print('4.', next(gen))
print('5.', next(gen))
print('6.', cvar.get())
```

The above prints:

```
1. initial
2. updated by callback
3. initial
4. updated at top level
5. updated by generator
6. updated by generator
```

Now add the following line after the creation of the generator:

```python
gen._context = contextvars.copy_context()
```

With that change, the script now outputs:

```
1. initial
2. initial
3. initial
4. initial
5. updated by generator
6. updated by top level
```

rhansen committed:
gh-99633: _contextvars: Call non-tstate variants

This will make it easier to refactor `_PyContext_Enter` and
`_PyContext_Exit` for a planned feature.

rhansen committed:
gh-124872: Change PyContext_WatchCallback to take PyObject

The PyContext struct is not intended to be public, and users of the
API don't need anything more specific than PyObject.  Also see
gh-78943.

rhansen committed:
gh-124872: Back up exception before calling PyContext_WatchCallback

I believe that the value of a simpler API (and defense against poorly
written callbacks) outweighs the cost of backing up and restoring the
thread's exception state.

rhansen committed:
gh-124872: Change PyContext_WatchCallback return type to void

The exception is ignored so change the return type from `int` to
`void` to discourage callbacks from raising an exception in the first
place.

rhansen committed:
gh-124872: Replace enter/exit events with "switched"

Users want to know when the current context switches to a different
context object.  Right now this happens when and only when a context
is entered or exited, so the enter and exit events are synonymous with
"switched".  However, if the changes proposed for gh-99633 are
implemented, the current context will also switch for reasons other
than context enter or exit.  Since users actually care about context
switches and not enter or exit, replace the enter and exit events with
a single switched event.

The former exit event was emitted just before exiting the context.
The new switched event is emitted after the context is exited to match
the semantics users expect of an event with a past-tense name.  If
users need the ability to clean up before the switch takes effect,
another event type can be added in the future.  It is not added here
because YAGNI.

I skipped 0 in the enum as a matter of practice.  Skipping 0 makes it
easier to troubleshoot when code forgets to set zeroed memory, and it
aligns with best practices for other tools (e.g.,
https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum).

rhansen committed:
gh-124872: Move PyThreadState to first argument for consistency

rhansen committed:
Merge branch 'gh-119333-exception' into HEAD

rhansen committed:
Merge branch 'gh-119333-notify' into HEAD

rhansen committed:
gh-99633: Inline independent coroutine check

The functions `_PyGen_ActivateContext` and `_PyGen_DeactivateContext`
are called every time a value or exception is sent to a coroutine.
These functions are no-ops for dependent coroutines (coroutines
without their own independent context stack).  Coroutines are
dependent by default, and the vast majority of performance-sensitive
coroutines are expected to be dependent, so move the check that
determines whether the coroutine is dependent or independent to an
inline function to speed up send calls.
